// @aws-amplify/graphql-relational-transformer/lib/graphql-has-many-transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HasManyTransformer = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const resolvers_1 = require("./resolvers");
const schema_1 = require("./schema");
const utils_1 = require("./utils");
const directiveName = 'hasMany';
const defaultLimit = 100;
const directiveDefinition = `
  directive @${directiveName}(indexName: String, fields: [String!], limit: Int = ${defaultLimit}) on FIELD_DEFINITION
`;
class HasManyTransformer extends graphql_transformer_core_1.TransformerPluginBase {
    constructor() {
        super('amplify-has-many-transformer', directiveDefinition);
        this.directiveList = [];
        this.field = (parent, definition, directive, context) => {
            const directiveWrapped = new graphql_transformer_core_1.DirectiveWrapper(directive);
            const args = directiveWrapped.getArguments({
                directiveName,
                object: parent,
                field: definition,
                directive,
                limit: defaultLimit,
            });
            validate(args, context);
            this.directiveList.push(args);
        };
        this.transformSchema = (ctx) => {
            const context = ctx;
            for (const config of this.directiveList) {
                config.relatedTypeIndex = utils_1.getRelatedTypeIndex(config, context, config.indexName);
                schema_1.ensureHasManyConnectionField(config, context);
                schema_1.extendTypeWithConnection(config, context);
            }
        };
        this.generateResolvers = (ctx) => {
            const context = ctx;
            for (const config of this.directiveList) {
                resolvers_1.updateTableForConnection(config, context);
                resolvers_1.makeQueryConnectionWithKeyResolver(config, context);
            }
        };
    }
}
exports.HasManyTransformer = HasManyTransformer;
function validate(config, ctx) {
    const { field } = config;
    utils_1.ensureFieldsArray(config);
    utils_1.validateModelDirective(config);
    if (!graphql_transformer_common_1.isListType(field.type)) {
        throw new graphql_transformer_core_1.InvalidDirectiveError(`@${directiveName} must be used with a list. Use @hasOne for non-list types.`);
    }
    config.fieldNodes = utils_1.getFieldsNodes(config, ctx);
    config.relatedType = utils_1.getRelatedType(config, ctx);
    config.connectionFields = [];
    utils_1.validateRelatedModelDirective(config);
}
