// @aws-amplify/graphql-searchable-transformer/lib/cdk/create-searchable-domain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSearchableDomainRole = exports.createSearchableDomain = void 0;
const aws_ec2_1 = require("@aws-cdk/aws-ec2");
const aws_elasticsearch_1 = require("@aws-cdk/aws-elasticsearch");
const aws_iam_1 = require("@aws-cdk/aws-iam");
const core_1 = require("@aws-cdk/core");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const createSearchableDomain = (stack, parameterMap, apiId) => {
    var _a, _b, _c;
    const { OpenSearchEBSVolumeGB, OpenSearchInstanceType, OpenSearchInstanceCount } = graphql_transformer_common_1.ResourceConstants.PARAMETERS;
    const { OpenSearchDomainLogicalID } = graphql_transformer_common_1.ResourceConstants.RESOURCES;
    const { HasEnvironmentParameter } = graphql_transformer_common_1.ResourceConstants.CONDITIONS;
    const domain = new aws_elasticsearch_1.Domain(stack, OpenSearchDomainLogicalID, {
        version: { version: '7.10' },
        ebs: {
            enabled: true,
            volumeType: aws_ec2_1.EbsDeviceVolumeType.GP2,
            volumeSize: (_a = parameterMap.get(OpenSearchEBSVolumeGB)) === null || _a === void 0 ? void 0 : _a.valueAsNumber,
        },
        zoneAwareness: {
            enabled: false,
        },
        domainName: core_1.Fn.conditionIf(HasEnvironmentParameter, core_1.Fn.ref('AWS::NoValue'), 'd' + apiId).toString(),
    });
    domain.node.defaultChild.elasticsearchClusterConfig = {
        instanceCount: (_b = parameterMap.get(OpenSearchInstanceCount)) === null || _b === void 0 ? void 0 : _b.valueAsNumber,
        instanceType: (_c = parameterMap.get(OpenSearchInstanceType)) === null || _c === void 0 ? void 0 : _c.valueAsString,
    };
    return domain;
};
exports.createSearchableDomain = createSearchableDomain;
const createSearchableDomainRole = (context, stack, parameterMap) => {
    const { OpenSearchAccessIAMRoleLogicalID } = graphql_transformer_common_1.ResourceConstants.RESOURCES;
    const { OpenSearchAccessIAMRoleName } = graphql_transformer_common_1.ResourceConstants.PARAMETERS;
    return new aws_iam_1.Role(stack, OpenSearchAccessIAMRoleLogicalID, {
        assumedBy: new aws_iam_1.ServicePrincipal('appsync.amazonaws.com'),
        roleName: context.resourceHelper.generateIAMRoleName(parameterMap.get(OpenSearchAccessIAMRoleName).valueAsString),
    });
};
exports.createSearchableDomainRole = createSearchableDomainRole;
