// amplify-category-geo/lib/service-stacks/mapStack.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapStack = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const location = __importStar(require("@aws-cdk/aws-location"));
const iam = __importStar(require("@aws-cdk/aws-iam"));
const core_1 = require("@aws-cdk/core");
const resourceParams_1 = require("../service-utils/resourceParams");
const baseStack_1 = require("./baseStack");
class MapStack extends baseStack_1.BaseStack {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props;
        this.accessType = this.props.accessType;
        this.parameters = this.constructInputParameters([
            'authRoleName',
            'unauthRoleName',
            'mapName',
            'mapStyle',
            'pricingPlan',
            'env',
            'isDefault'
        ]);
        this.resources = this.constructResources();
        new cdk.CfnOutput(this, 'Name', {
            value: this.resources.get('map').ref
        });
        new cdk.CfnOutput(this, 'Style', {
            value: this.parameters.get('mapStyle').valueAsString
        });
    }
    constructResources() {
        let resourcesMap = new Map();
        const mapResource = this.constructMapResource();
        resourcesMap.set('map', mapResource);
        const mapPolicyResource = this.constructMapPolicyResource(mapResource);
        resourcesMap.set('mapPolicy', mapPolicyResource);
        return resourcesMap;
    }
    constructMapResource() {
        return new location.CfnMap(this, 'Map', {
            mapName: core_1.Fn.join('-', [
                this.parameters.get('mapName').valueAsString,
                this.parameters.get('env').valueAsString
            ]),
            configuration: {
                style: this.parameters.get('mapStyle').valueAsString
            },
            pricingPlan: this.parameters.get('pricingPlan').valueAsString
        });
    }
    constructMapPolicyResource(mapResource) {
        let policy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    effect: iam.Effect.ALLOW,
                    actions: [
                        "geo:GetMapStyleDescriptor",
                        "geo:GetMapGlyphs",
                        "geo:GetMapSprites",
                        "geo:GetMapTile"
                    ],
                    resources: [mapResource.getAtt('MapArn').toString()],
                })
            ],
        });
        let cognitoRoles = new Array();
        cognitoRoles.push(this.parameters.get('authRoleName').valueAsString);
        if (this.accessType == resourceParams_1.AccessType.AuthorizedAndGuestUsers) {
            cognitoRoles.push(this.parameters.get('unauthRoleName').valueAsString);
        }
        return new iam.CfnPolicy(this, 'MapPolicy', {
            policyName: cdk.Fn.join('-', [
                this.parameters.get('mapName').valueAsString,
                this.parameters.get('env').valueAsString,
                'Policy'
            ]),
            roles: cognitoRoles,
            policyDocument: policy
        });
    }
}
exports.MapStack = MapStack;
