// amplify-util-mock/lib/utils/config-override.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigOverrideManager = void 0;
const index_1 = require("./index");
class ConfigOverrideManager {
    constructor(context) {
        this.overrides = {};
        context.amplify.addCleanUpTask(async (context) => {
            await this.restoreFrontendExports(context);
        });
    }
    addOverride(category, override) {
        this.overrides[category] = override;
    }
    async generateOverriddenFrontendExports(context) {
        const meta = await index_1.getAmplifyMeta(context);
        await context.amplify.onCategoryOutputsChange(context, null, {
            ...meta,
            ...this.overrides,
        });
    }
    async restoreFrontendExports(context) {
        const meta = await index_1.getAmplifyMeta(context);
        await context.amplify.onCategoryOutputsChange(context, null, meta);
    }
    static getInstance(context) {
        if (!ConfigOverrideManager.instance) {
            ConfigOverrideManager.instance = new ConfigOverrideManager(context);
        }
        return ConfigOverrideManager.instance;
    }
}
exports.ConfigOverrideManager = ConfigOverrideManager;
ConfigOverrideManager.instance = null;
