// amplify-category-geo/lib/service-utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceName = exports.provider = exports.parametersFileName = exports.choosePricingPlan = exports.chooseServiceMessageRemove = exports.chooseServiceMessageUpdate = exports.chooseServiceMessageAdd = exports.previewBanner = exports.apiDocs = void 0;
exports.apiDocs = {
    mapStyles: "https://docs.aws.amazon.com/location-maps/latest/APIReference/API_MapConfiguration.html",
    pricingPlan: "https://aws.amazon.com/location/pricing/",
    dataSourceUsage: "https://docs.aws.amazon.com/location-places/latest/APIReference/API_DataSourceConfiguration.html"
};
exports.previewBanner = 'Amplify Geo category is in developer preview and not intended for production use at this time.';
exports.chooseServiceMessageAdd = 'Select which capability you want to add:';
exports.chooseServiceMessageUpdate = 'Select which capability you want to update:';
exports.chooseServiceMessageRemove = 'Select which capability you want to remove:';
exports.choosePricingPlan = `The following choices determine the pricing plan for Geo resources. Learn more at ${exports.apiDocs.pricingPlan}`;
exports.parametersFileName = 'parameters.json';
exports.provider = 'awscloudformation';
var ServiceName;
(function (ServiceName) {
    ServiceName["Map"] = "Map";
    ServiceName["PlaceIndex"] = "PlaceIndex";
    ServiceName["GeofenceCollection"] = "GeofenceCollection";
    ServiceName["Tracker"] = "Tracker";
})(ServiceName = exports.ServiceName || (exports.ServiceName = {}));
