// amplify-category-geo/lib/service-utils/mapParams.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMapStyleComponents = exports.getGeoMapStyle = exports.convertToCompleteMapParams = exports.isCompleteMapParams = exports.MapStyle = exports.HereMapStyleType = exports.EsriMapStyleType = void 0;
const resourceParams_1 = require("./resourceParams");
const lodash_1 = __importDefault(require("lodash"));
var EsriMapStyleType;
(function (EsriMapStyleType) {
    EsriMapStyleType["Navigation"] = "Navigation";
    EsriMapStyleType["Streets"] = "Streets";
    EsriMapStyleType["Topographic"] = "Topographic";
    EsriMapStyleType["DarkGrayCanvas"] = "DarkGrayCanvas";
    EsriMapStyleType["LightGrayCanvas"] = "LightGrayCanvas";
})(EsriMapStyleType = exports.EsriMapStyleType || (exports.EsriMapStyleType = {}));
var HereMapStyleType;
(function (HereMapStyleType) {
    HereMapStyleType["Berlin"] = "Berlin";
})(HereMapStyleType = exports.HereMapStyleType || (exports.HereMapStyleType = {}));
var MapStyle;
(function (MapStyle) {
    MapStyle["VectorEsriNavigation"] = "VectorEsriNavigation";
    MapStyle["VectorEsriStreets"] = "VectorEsriStreets";
    MapStyle["VectorEsriTopographic"] = "VectorEsriTopographic";
    MapStyle["VectorEsriDarkGrayCanvas"] = "VectorEsriDarkGrayCanvas";
    MapStyle["VectorEsriLightGrayCanvas"] = "VectorEsriLightGrayCanvas";
    MapStyle["VectorHereBerlin"] = "VectorHereBerlin";
})(MapStyle = exports.MapStyle || (exports.MapStyle = {}));
const isCompleteMapParams = (partial) => {
    const requiredFields = ['providerContext', 'name', 'mapStyleType', 'dataProvider', 'accessType', 'isDefault'];
    const missingField = requiredFields.find(field => !lodash_1.default.keys(partial).includes(field));
    return !missingField;
};
exports.isCompleteMapParams = isCompleteMapParams;
const convertToCompleteMapParams = (partial) => {
    if (exports.isCompleteMapParams(partial)) {
        return partial;
    }
    throw new Error('Partial<MapParameters> does not satisfy MapParameters');
};
exports.convertToCompleteMapParams = convertToCompleteMapParams;
const getGeoMapStyle = (dataProvider, mapStyleType) => {
    if (dataProvider === resourceParams_1.DataProvider.Here && mapStyleType === HereMapStyleType.Berlin) {
        return MapStyle.VectorHereBerlin;
    }
    return `Vector${dataProvider}${mapStyleType}`;
};
exports.getGeoMapStyle = getGeoMapStyle;
const getMapStyleComponents = (mapStyle) => {
    switch (mapStyle) {
        case MapStyle.VectorEsriDarkGrayCanvas:
            return { dataProvider: resourceParams_1.DataProvider.Esri, mapStyleType: EsriMapStyleType.DarkGrayCanvas };
        case MapStyle.VectorEsriLightGrayCanvas:
            return { dataProvider: resourceParams_1.DataProvider.Esri, mapStyleType: EsriMapStyleType.LightGrayCanvas };
        case MapStyle.VectorEsriNavigation:
            return { dataProvider: resourceParams_1.DataProvider.Esri, mapStyleType: EsriMapStyleType.Navigation };
        case MapStyle.VectorEsriStreets:
            return { dataProvider: resourceParams_1.DataProvider.Esri, mapStyleType: EsriMapStyleType.Streets };
        case MapStyle.VectorEsriTopographic:
            return { dataProvider: resourceParams_1.DataProvider.Esri, mapStyleType: EsriMapStyleType.Topographic };
        case MapStyle.VectorHereBerlin:
            return { dataProvider: resourceParams_1.DataProvider.Here, mapStyleType: HereMapStyleType.Berlin };
        default:
            throw new Error(`Invalid map style ${mapStyle}`);
    }
};
exports.getMapStyleComponents = getMapStyleComponents;
