// amplify-category-geo/lib/service-walkthroughs/placeIndexWalkthrough.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateDefaultPlaceIndexWalkthrough = exports.updatePlaceIndexWalkthrough = exports.placeIndexDataStorageWalkthrough = exports.placeIndexAdvancedWalkthrough = exports.placeIndexNameWalkthrough = exports.createPlaceIndexWalkthrough = void 0;
const uuid_1 = require("uuid");
const resourceUtils_1 = require("../service-utils/resourceUtils");
const placeIndexParams_1 = require("../service-utils/placeIndexParams");
const constants_1 = require("../service-utils/constants");
const placeIndexUtils_1 = require("../service-utils/placeIndexUtils");
const resourceUtils_2 = require("../service-utils/resourceUtils");
const resourceWalkthrough_1 = require("./resourceWalkthrough");
const resourceParams_1 = require("../service-utils/resourceParams");
const amplify_prompts_1 = require("amplify-prompts");
const searchServiceFriendlyName = resourceWalkthrough_1.getServiceFriendlyName(constants_1.ServiceName.PlaceIndex);
const createPlaceIndexWalkthrough = async (context, parameters) => {
    parameters = resourceUtils_1.merge(parameters, await exports.placeIndexNameWalkthrough(context));
    parameters = resourceUtils_1.merge(parameters, await resourceWalkthrough_1.resourceAccessWalkthrough(parameters, constants_1.ServiceName.PlaceIndex));
    if (!(await resourceUtils_2.geoServiceExists(constants_1.ServiceName.Map)) && !(await resourceUtils_2.geoServiceExists(constants_1.ServiceName.PlaceIndex))) {
        parameters = resourceUtils_1.merge(parameters, await resourceWalkthrough_1.pricingPlanWalkthrough(context, parameters));
    }
    parameters = resourceUtils_1.merge(parameters, await exports.placeIndexAdvancedWalkthrough(context, parameters));
    const currentPlaceIndexResources = await resourceUtils_2.getGeoServiceMeta(constants_1.ServiceName.PlaceIndex);
    if (currentPlaceIndexResources && Object.keys(currentPlaceIndexResources).length > 0) {
        parameters.isDefault = await amplify_prompts_1.prompter.yesOrNo(resourceWalkthrough_1.defaultResourceQuestion(constants_1.ServiceName.PlaceIndex), true);
    }
    else {
        parameters.isDefault = true;
    }
    return parameters;
};
exports.createPlaceIndexWalkthrough = createPlaceIndexWalkthrough;
const placeIndexNameWalkthrough = async (context) => {
    let indexName;
    while (!indexName) {
        const [shortId] = uuid_1.v4().split('-');
        const indexNameInput = await amplify_prompts_1.prompter.input('Provide a name for the location search index (place index):', { validate: amplify_prompts_1.alphanumeric(), initial: `placeindex${shortId}` });
        if (await resourceUtils_2.checkGeoResourceExists(indexNameInput)) {
            amplify_prompts_1.printer.info(`Location search index ${indexNameInput} already exists. Choose another name.`);
        }
        else
            indexName = indexNameInput;
    }
    return { name: indexName };
};
exports.placeIndexNameWalkthrough = placeIndexNameWalkthrough;
const placeIndexAdvancedWalkthrough = async (context, parameters) => {
    const includePricingPlan = false;
    const currentPricingPlan = parameters.pricingPlan ? parameters.pricingPlan : await resourceUtils_2.getGeoPricingPlan();
    const advancedSettingOptions = ['Search data provider (default: Esri)'];
    advancedSettingOptions.push('Search result storage location (default: no result storage)');
    if (includePricingPlan) {
        advancedSettingOptions.push(`Search pricing plan (current: ${currentPricingPlan})`);
    }
    amplify_prompts_1.printer.info('Available advanced settings:');
    amplify_prompts_1.formatter.list(advancedSettingOptions);
    amplify_prompts_1.printer.blankLine();
    if (await amplify_prompts_1.prompter.yesOrNo('Do you want to configure advanced settings?', false)) {
        parameters = resourceUtils_1.merge(parameters, await resourceWalkthrough_1.dataProviderWalkthrough(parameters, constants_1.ServiceName.PlaceIndex));
        if (includePricingPlan) {
            parameters = resourceUtils_1.merge(parameters, await resourceWalkthrough_1.pricingPlanWalkthrough(context, parameters));
        }
        else {
            parameters.pricingPlan = currentPricingPlan;
        }
        if (parameters.pricingPlan === resourceParams_1.PricingPlan.RequestBasedUsage) {
            parameters = resourceUtils_1.merge(parameters, await exports.placeIndexDataStorageWalkthrough(context, parameters));
        }
        else {
            parameters.dataSourceIntendedUse = placeIndexParams_1.DataSourceIntendedUse.SingleUse;
        }
    }
    else {
        parameters.dataProvider = resourceParams_1.DataProvider.Esri;
        parameters.dataSourceIntendedUse = placeIndexParams_1.DataSourceIntendedUse.SingleUse;
        parameters.pricingPlan = currentPricingPlan;
    }
    return parameters;
};
exports.placeIndexAdvancedWalkthrough = placeIndexAdvancedWalkthrough;
const placeIndexDataStorageWalkthrough = async (context, parameters) => {
    const areResultsStored = await amplify_prompts_1.prompter.yesOrNo(`Do you want to cache or store the results of search operations? Refer ${constants_1.apiDocs.dataSourceUsage}`, parameters.dataSourceIntendedUse === placeIndexParams_1.DataSourceIntendedUse.Storage);
    const intendedUse = areResultsStored ? placeIndexParams_1.DataSourceIntendedUse.Storage : placeIndexParams_1.DataSourceIntendedUse.SingleUse;
    return { dataSourceIntendedUse: intendedUse };
};
exports.placeIndexDataStorageWalkthrough = placeIndexDataStorageWalkthrough;
const updatePlaceIndexWalkthrough = async (context, parameters, resourceToUpdate) => {
    const indexResources = (await context.amplify.getResourceStatus()).allResources
        .filter(resource => resource.service === constants_1.ServiceName.PlaceIndex);
    if (indexResources.length === 0) {
        amplify_prompts_1.printer.error(`No ${searchServiceFriendlyName} resource to update. Use "amplify add geo" to create a new ${searchServiceFriendlyName}.`);
        return parameters;
    }
    const indexResourceNames = indexResources.map(resource => resource.resourceName);
    if (resourceToUpdate) {
        if (!indexResourceNames.includes(resourceToUpdate)) {
            amplify_prompts_1.printer.error(`No ${searchServiceFriendlyName} named ${resourceToUpdate} exists in the project.`);
            return parameters;
        }
    }
    else {
        resourceToUpdate = await amplify_prompts_1.prompter.pick(`Select the ${searchServiceFriendlyName} you want to update`, indexResourceNames);
    }
    parameters.name = resourceToUpdate;
    parameters = resourceUtils_1.merge(parameters, await placeIndexUtils_1.getCurrentPlaceIndexParameters(resourceToUpdate));
    parameters.accessType = (await resourceWalkthrough_1.resourceAccessWalkthrough(parameters, constants_1.ServiceName.PlaceIndex)).accessType;
    const otherIndexResources = indexResourceNames.filter(indexResourceName => indexResourceName != resourceToUpdate);
    if (otherIndexResources.length > 0) {
        const isDefault = await amplify_prompts_1.prompter.yesOrNo(resourceWalkthrough_1.defaultResourceQuestion(constants_1.ServiceName.PlaceIndex), true);
        if (parameters.isDefault && !isDefault) {
            await exports.updateDefaultPlaceIndexWalkthrough(context, resourceToUpdate, otherIndexResources);
        }
        parameters.isDefault = isDefault;
    }
    else {
        parameters.isDefault = true;
    }
    return parameters;
};
exports.updatePlaceIndexWalkthrough = updatePlaceIndexWalkthrough;
const updateDefaultPlaceIndexWalkthrough = async (context, currentDefault, availablePlaceIndices) => {
    if (!availablePlaceIndices) {
        availablePlaceIndices = (await context.amplify.getResourceStatus()).allResources
            .filter(resource => resource.service === constants_1.ServiceName.PlaceIndex)
            .map(resource => resource.resourceName);
    }
    const otherIndexResources = availablePlaceIndices.filter(indexResourceName => indexResourceName != currentDefault);
    if ((otherIndexResources === null || otherIndexResources === void 0 ? void 0 : otherIndexResources.length) > 0) {
        const defaultIndexName = await amplify_prompts_1.prompter.pick(`Select the ${searchServiceFriendlyName} you want to set as default:`, otherIndexResources);
        await resourceUtils_2.updateDefaultResource(context, constants_1.ServiceName.PlaceIndex, defaultIndexName);
    }
    return currentDefault;
};
exports.updateDefaultPlaceIndexWalkthrough = updateDefaultPlaceIndexWalkthrough;
