// @aws-amplify/graphql-transformer-core/lib/utils/directive-wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectiveWrapper = void 0;
const graphql_1 = require("graphql");
const lodash_1 = require("lodash");
class ArgumentWrapper {
    constructor(argument) {
        this.serialize = () => {
            return {
                kind: 'Argument',
                name: this.name,
                value: this.value,
            };
        };
        this.name = argument.name;
        this.value = argument.value;
    }
}
class DirectiveWrapper {
    constructor(node) {
        var _a;
        this.arguments = [];
        this.serialize = () => {
            return {
                kind: 'Directive',
                name: this.name,
                arguments: this.arguments.map(arg => arg.serialize()),
            };
        };
        this.getArguments = (defaultValue) => {
            const argValues = this.arguments.reduce((acc, arg) => ({
                ...acc,
                [arg.name.value]: graphql_1.valueFromASTUntyped(arg.value),
            }), {});
            return lodash_1.merge(defaultValue, argValues);
        };
        this.name = node.name;
        this.arguments = ((_a = node.arguments) !== null && _a !== void 0 ? _a : []).map(arg => new ArgumentWrapper(arg));
        this.location = this.location;
    }
}
exports.DirectiveWrapper = DirectiveWrapper;
