// @aws-amplify/graphql-http-transformer/lib/graphql-http-transformer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpTransformer = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const cdk = __importStar(require("@aws-cdk/core"));
const graphql_1 = require("graphql");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const graphql_mapping_template_1 = require("graphql-mapping-template");
const SPLIT_URL_REGEX = /(http(s)?:\/\/|www\.)|(\/.*)/g;
const URL_REGEX = /(http(s)?:\/\/)|(\/.*)/g;
const VALID_PROTOCOLS_REGEX = /^http(s)?:\/\//;
const HTTP_DIRECTIVE_STACK = 'HttpDirectiveStack';
const RESOLVER_VERSION = '2018-05-29';
const directiveDefinition = `
  directive @http(method: HttpMethod = GET, url: String!, headers: [HttpHeader] = []) on FIELD_DEFINITION
  enum HttpMethod {
    GET
    POST
    PUT
    DELETE
    PATCH
  }
  input HttpHeader {
    key: String
    value: String
  }
`;
class HttpTransformer extends graphql_transformer_core_1.TransformerPluginBase {
    constructor() {
        super('amplify-http-transformer', directiveDefinition);
        this.directiveList = [];
        this.field = (parent, definition, directive, context) => {
            var _a, _b;
            const directiveWrapped = new graphql_transformer_core_1.DirectiveWrapper(directive);
            const args = directiveWrapped.getArguments({
                method: 'GET',
                path: '',
                origin: '',
                queryAndBodyArgs: definition.arguments,
                resolverTypeName: parent.name.value,
                resolverFieldName: definition.name.value,
                supportsBody: false,
            });
            if (!VALID_PROTOCOLS_REGEX.test(args.url)) {
                throw new graphql_transformer_core_1.TransformerContractError(`@http directive at location ${(_a = directive === null || directive === void 0 ? void 0 : directive.loc) === null || _a === void 0 ? void 0 : _a.start} requires a url parameter that begins with http:// or https://.`);
            }
            args.origin = args.url.replace(URL_REGEX, '$1');
            args.path = (_b = args.url.split(SPLIT_URL_REGEX).slice(-2, -1)[0]) !== null && _b !== void 0 ? _b : '/';
            args.supportsBody = args.method === 'POST' || args.method === 'PUT' || args.method === 'PATCH';
            if (!args.headers) {
                args.headers = [];
            }
            else if (!Array.isArray(args.headers)) {
                args.headers = [args.headers];
            }
            const newFieldArgsArray = [];
            let params = args.path.match(/:\w+/g);
            if (params) {
                params = params.map(p => p.replace(':', ''));
                args.queryAndBodyArgs = args.queryAndBodyArgs.filter(arg => {
                    return graphql_transformer_common_1.isScalar(arg.type) && !params.includes(arg.name.value);
                });
                args.path = args.path.replace(/:\w+/g, (str) => {
                    return `\$\{ctx.args.params.${str.replace(':', '')}\}`;
                });
                const urlParamInputObject = makeUrlParamInputObject(args, params);
                context.output.addInput(urlParamInputObject);
                newFieldArgsArray.push(makeHttpArgument('params', urlParamInputObject, true));
            }
            if (args.queryAndBodyArgs.length > 0) {
                const name = graphql_transformer_common_1.ModelResourceIDs.HttpQueryInputObjectName(parent.name.value, definition.name.value);
                const queryInputObject = makeHttpInputObject(name, args.queryAndBodyArgs, args.method !== 'GET');
                const makeNonNull = queryInputObject.fields.filter(a => a.type.kind === graphql_1.Kind.NON_NULL_TYPE).length > 0;
                context.output.addInput(queryInputObject);
                newFieldArgsArray.push(makeHttpArgument('query', queryInputObject, makeNonNull));
                if (args.supportsBody) {
                    const name = graphql_transformer_common_1.ModelResourceIDs.HttpBodyInputObjectName(parent.name.value, definition.name.value);
                    const bodyInputObject = makeHttpInputObject(name, args.queryAndBodyArgs, true);
                    context.output.addInput(bodyInputObject);
                    newFieldArgsArray.push(makeHttpArgument('body', bodyInputObject, makeNonNull));
                }
            }
            if (newFieldArgsArray.length > 0) {
                const updatedField = {
                    ...definition,
                    arguments: newFieldArgsArray,
                };
                const mostRecentParent = context.output.getType(parent.name.value);
                let updatedFieldsInParent = mostRecentParent.fields.filter(f => f.name.value !== definition.name.value);
                updatedFieldsInParent.push(updatedField);
                const updatedParentType = {
                    ...mostRecentParent,
                    fields: updatedFieldsInParent,
                };
                context.output.putType(updatedParentType);
            }
            this.directiveList.push(args);
        };
        this.generateResolvers = (context) => {
            if (this.directiveList.length === 0) {
                return;
            }
            const stack = context.stackManager.createStack(HTTP_DIRECTIVE_STACK);
            const env = context.stackManager.getParameter(graphql_transformer_common_1.ResourceConstants.PARAMETERS.Env);
            const region = stack.region;
            stack.templateOptions.templateFormatVersion = '2010-09-09';
            stack.templateOptions.description = 'An auto-generated nested stack for the @http directive.';
            this.directiveList.forEach(directive => {
                const dataSourceId = graphql_transformer_common_1.HttpResourceIDs.HttpDataSourceID(directive.origin);
                if (context.api.host.getDataSource(dataSourceId) === undefined) {
                    context.api.host.addHttpDataSource(dataSourceId, replaceEnvAndRegion(env, region, directive.origin), {}, stack);
                }
                createResolver(stack, dataSourceId, context, directive);
            });
        };
    }
}
exports.HttpTransformer = HttpTransformer;
function createResolver(stack, dataSourceId, context, config) {
    const env = context.stackManager.getParameter(graphql_transformer_common_1.ResourceConstants.PARAMETERS.Env);
    const region = stack.region;
    const { method, supportsBody } = config;
    const reqCompoundExpr = [];
    const requestParams = { headers: graphql_mapping_template_1.ref('util.toJson($headers)') };
    const parsedHeaders = config.headers.map(header => graphql_mapping_template_1.qref(`$headers.put("${header.key}", "${header.value}")`));
    if (method !== 'DELETE') {
        requestParams.query = graphql_mapping_template_1.ref('util.toJson($ctx.args.query)');
    }
    if (supportsBody) {
        const nonNullArgs = config.queryAndBodyArgs.filter(arg => arg.type.kind === graphql_1.Kind.NON_NULL_TYPE);
        requestParams.body = graphql_mapping_template_1.ref('util.toJson($ctx.args.body)');
        if (nonNullArgs.length > 0) {
            reqCompoundExpr.push(graphql_mapping_template_1.compoundExpression([
                graphql_mapping_template_1.comment('START: Manually checking that all non-null arguments are provided either in the query or the body'),
                graphql_mapping_template_1.iff(graphql_mapping_template_1.or(nonNullArgs.map(arg => {
                    const name = arg.name.value;
                    return graphql_mapping_template_1.parens(graphql_mapping_template_1.and([graphql_mapping_template_1.raw(`!$ctx.args.body.${name}`), graphql_mapping_template_1.raw(`!$ctx.args.query.${name}`)]));
                })), graphql_mapping_template_1.ref('util.error("An argument you marked as Non-Null is not present in the query nor the body of your request."))')),
                graphql_mapping_template_1.comment('END: Manually checking that all non-null arguments are provided either in the query or the body'),
            ]));
        }
    }
    reqCompoundExpr.push(graphql_mapping_template_1.set(graphql_mapping_template_1.ref('headers'), graphql_mapping_template_1.ref('utils.http.copyHeaders($ctx.request.headers)')), graphql_mapping_template_1.qref('$headers.put("accept-encoding", "application/json")'));
    if (supportsBody) {
        reqCompoundExpr.push(graphql_mapping_template_1.qref('$headers.put("Content-Type", "application/json")'));
    }
    reqCompoundExpr.push(...parsedHeaders, graphql_mapping_template_1.obj({
        version: graphql_mapping_template_1.str(RESOLVER_VERSION),
        method: graphql_mapping_template_1.str(method),
        resourcePath: graphql_mapping_template_1.str(config.path),
        params: graphql_mapping_template_1.obj(requestParams),
    }));
    const requestTemplateString = replaceEnvAndRegion(env, region, graphql_mapping_template_1.printBlock('Create request')(graphql_mapping_template_1.compoundExpression(reqCompoundExpr)));
    const requestMappingTemplate = cdk.Token.isUnresolved(requestTemplateString)
        ? graphql_transformer_core_1.MappingTemplate.inlineTemplateFromString(requestTemplateString)
        : graphql_transformer_core_1.MappingTemplate.s3MappingTemplateFromString(requestTemplateString, `${config.resolverTypeName}.${config.resolverFieldName}.req.vtl`);
    return context.api.host.addResolver(config.resolverTypeName, config.resolverFieldName, requestMappingTemplate, graphql_transformer_core_1.MappingTemplate.s3MappingTemplateFromString(graphql_mapping_template_1.printBlock('Process response')(graphql_mapping_template_1.ifElse(supportsBody ? graphql_mapping_template_1.raw('$ctx.result.statusCode == 200 || $ctx.result.statusCode == 201') : graphql_mapping_template_1.raw('$ctx.result.statusCode == 200'), graphql_mapping_template_1.ifElse(graphql_mapping_template_1.ref('ctx.result.headers.get("Content-Type").toLowerCase().contains("xml")'), graphql_mapping_template_1.ref('utils.xml.toJsonString($ctx.result.body)'), graphql_mapping_template_1.ref('ctx.result.body')), graphql_mapping_template_1.ref('util.qr($util.appendError($ctx.result.body, $ctx.result.statusCode))'))), `${config.resolverTypeName}.${config.resolverFieldName}.res.vtl`), dataSourceId, undefined, stack);
}
function replaceEnvAndRegion(env, region, value) {
    const vars = {};
    if (value.includes('${env}')) {
        vars.env = env;
    }
    if (value.includes('${aws_region}')) {
        vars.aws_region = region;
    }
    if (Object.keys(vars).length === 0) {
        return value;
    }
    return cdk.Fn.sub(value, vars);
}
function makeUrlParamInputObject(directive, urlParams) {
    return {
        kind: 'InputObjectTypeDefinition',
        name: {
            kind: 'Name',
            value: graphql_transformer_common_1.ModelResourceIDs.UrlParamsInputObjectName(directive.resolverTypeName, directive.resolverFieldName),
        },
        fields: urlParams.map(param => {
            return graphql_transformer_common_1.makeInputValueDefinition(param, graphql_transformer_common_1.makeNonNullType(graphql_transformer_common_1.makeNamedType('String')));
        }),
        directives: [],
    };
}
function makeHttpArgument(name, inputType, makeNonNull) {
    const type = makeNonNull ? graphql_transformer_common_1.makeNonNullType(graphql_transformer_common_1.makeNamedType(inputType.name.value)) : graphql_transformer_common_1.makeNamedType(inputType.name.value);
    return graphql_transformer_common_1.makeInputValueDefinition(name, type);
}
function makeHttpInputObject(name, argArray, makeNonNull) {
    const fields = makeNonNull
        ? argArray.map((arg) => {
            return {
                ...arg,
                type: graphql_transformer_common_1.unwrapNonNull(arg.type),
            };
        })
        : argArray;
    return {
        kind: 'InputObjectTypeDefinition',
        name: {
            kind: 'Name',
            value: name,
        },
        fields,
        directives: [],
    };
}
