// amplify-provider-awscloudformation/lib/iterative-deployment/deployment-manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeploymentManager = exports.DeploymentError = void 0;
const aws = __importStar(require("aws-sdk"));
const state_machine_1 = require("./state-machine");
const stack_event_monitor_1 = require("./stack-event-monitor");
const helpers_1 = require("./helpers");
const stack_progress_printer_1 = require("./stack-progress-printer");
const assert_1 = __importDefault(require("assert"));
const configuration_manager_1 = require("../configuration-manager");
const xstate_1 = require("xstate");
const lodash_throttle_1 = __importDefault(require("lodash.throttle"));
const aws_logger_1 = require("../utils/aws-logger");
class DeploymentError extends Error {
    constructor(errors) {
        super('There was an error while deploying changes.');
        this.name = `DeploymentError`;
        const stackTrace = [];
        for (const err of errors) {
            stackTrace.push(`Index: ${err.currentIndex} State: ${err.stateValue} Message: ${err.error.message}`);
        }
        this.message = JSON.stringify(stackTrace);
    }
}
exports.DeploymentError = DeploymentError;
class DeploymentManager {
    constructor(creds, region, deploymentBucket, spinner, printer = new stack_progress_printer_1.StackProgressPrinter(), options = {}) {
        this.region = region;
        this.deploymentBucket = deploymentBucket;
        this.spinner = spinner;
        this.printer = printer;
        this.deployment = [];
        this.deploy = async (deploymentStateManager) => {
            this.deploymentStateManager = deploymentStateManager;
            const deploymentTemplates = this.deployment.reduce((acc, step) => {
                acc.add(step.deployment.stackTemplatePath);
                acc.add(step.rollback.stackTemplatePath);
                return acc;
            }, new Set());
            await Promise.all(Array.from(deploymentTemplates.values()).map(path => this.ensureTemplateExists(path)));
            const fns = {
                deployFn: this.doDeploy,
                deploymentWaitFn: this.waitForDeployment,
                rollbackFn: this.rollBackStack,
                tableReadyWaitFn: this.waitForIndices,
                rollbackWaitFn: this.waitForDeployment,
                stackEventPollFn: this.stackPollFn,
                startRollbackFn: this.startRollBackFn,
            };
            const machine = state_machine_1.createDeploymentMachine({
                currentIndex: -1,
                deploymentBucket: this.deploymentBucket,
                region: this.region,
                stacks: this.deployment,
            }, fns);
            let maxDeployed = 0;
            return new Promise(async (resolve, reject) => {
                const service = xstate_1.interpret(machine)
                    .onTransition(async (state) => {
                    if (state.changed) {
                        this.spinner.render();
                        maxDeployed = Math.max(maxDeployed, state.context.currentIndex + 1);
                        if (state.matches('idle')) {
                            this.logSpinnerMessage('deploy', `Starting deployment`);
                        }
                        else if (state.matches('deploy.waitForTablesToBeReady')) {
                            this.logSpinnerMessage('deploy', `Waiting for DynamoDB indices to be ready`);
                        }
                        else if (state.matches('deploy')) {
                            this.logSpinnerMessage('deploy', `Deploying (${maxDeployed} of ${state.context.stacks.length})`);
                        }
                        else if (state.matches('rollback')) {
                            this.logSpinnerMessage('deploy', `Rolling back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
                        }
                        else if (state.matches('deployed')) {
                            this.logger('DeploymentManager', [{ spinner: 'Deployed' }])();
                            this.spinner.succeed(`Deployed`);
                        }
                    }
                    switch (state.value) {
                        case 'deployed':
                            return resolve();
                        case 'rolledBack':
                        case 'failed':
                            const deploymentErrors = new DeploymentError(state.context.errors);
                            this.logger('DeploymentManager', [{ stateValue: state.value }])(deploymentErrors);
                            return reject(deploymentErrors);
                        default:
                    }
                })
                    .start();
                service.send({ type: 'DEPLOY' });
            });
        };
        this.rollback = async (deploymentStateManager) => {
            this.deploymentStateManager = deploymentStateManager;
            let currentStepIndex = this.deploymentStateManager.getStatus().currentStepIndex;
            let maxDeployed = currentStepIndex + 1;
            const deploymentTemplates = this.deployment.reduce((acc, step) => {
                acc.add(step.rollback.stackTemplatePath);
                return acc;
            }, new Set());
            await Promise.all(Array.from(deploymentTemplates.values()).map(path => this.ensureTemplateExists(path)));
            const fns = {
                preRollbackTableCheck: this.preRollbackTableCheck,
                rollbackFn: this.rollBackStack,
                tableReadyWaitFn: this.waitForIndices,
                rollbackWaitFn: this.waitForDeployment,
                stackEventPollFn: this.stackPollFn,
                startRollbackFn: this.startRollBackFn,
            };
            const machine = state_machine_1.createRollbackDeploymentMachine({
                currentIndex: maxDeployed,
                previousDeploymentIndex: currentStepIndex,
                deploymentBucket: this.deploymentBucket,
                region: this.region,
                stacks: this.deployment,
            }, fns);
            return new Promise(async (resolve, reject) => {
                const service = xstate_1.interpret(machine)
                    .onTransition(async (state) => {
                    if (state.changed) {
                        this.spinner.render();
                        if (state.matches('idle')) {
                            this.logSpinnerMessage('rollback', `Starting rollback`);
                        }
                        else if (state.matches('preRollback')) {
                            this.logSpinnerMessage('rollback', `Waiting for previous deployment to finish`);
                        }
                        else if (state.matches('rollback.waitForTablesToBeReady')) {
                            this.logSpinnerMessage('rollback', `Waiting for DynamoDB indices to be ready`);
                        }
                        else if (state.matches('rollback')) {
                            this.logSpinnerMessage('rollback', `Rolling back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
                        }
                        else if (state.matches('rolledBack')) {
                            this.spinner.succeed(`Rolled back succesfully`);
                            this.logger('rollback', [{ spinner: 'Rolled back successfully' }])();
                        }
                    }
                    switch (state.value) {
                        case 'rolledBack':
                            return resolve();
                        case 'failed':
                            const deploymentErrors = new DeploymentError(state.context.errors);
                            this.logger('DeploymentManager', [{ stateValue: state.value }])(deploymentErrors);
                            return reject(deploymentErrors);
                        default:
                    }
                })
                    .start();
                service.send({ type: 'ROLLBACK' });
            });
        };
        this.addStep = (deploymentStep) => {
            const deploymentStackTemplateUrl = helpers_1.getHttpUrl(deploymentStep.deployment.stackTemplatePathOrUrl, this.deploymentBucket);
            const deploymentStackTemplatePath = helpers_1.getBucketKey(deploymentStep.deployment.stackTemplatePathOrUrl, this.deploymentBucket);
            const rollbackStackTemplateUrl = helpers_1.getHttpUrl(deploymentStep.rollback.stackTemplatePathOrUrl, this.deploymentBucket);
            const rollbackStackTemplatePath = helpers_1.getBucketKey(deploymentStep.rollback.stackTemplatePathOrUrl, this.deploymentBucket);
            this.deployment.push({
                deployment: {
                    ...deploymentStep.deployment,
                    stackTemplatePath: deploymentStackTemplatePath,
                    stackTemplateUrl: deploymentStackTemplateUrl,
                    region: this.region,
                    clientRequestToken: deploymentStep.deployment.clientRequestToken
                        ? `deploy-${deploymentStep.deployment.clientRequestToken}`
                        : undefined,
                },
                rollback: {
                    ...deploymentStep.rollback,
                    stackTemplatePath: rollbackStackTemplatePath,
                    stackTemplateUrl: rollbackStackTemplateUrl,
                    region: this.region,
                    clientRequestToken: deploymentStep.rollback.clientRequestToken
                        ? `rollback-${deploymentStep.rollback.clientRequestToken}`
                        : undefined,
                },
            });
        };
        this.addRollbackStep = (rollbackStep) => {
            const rollbackStackTemplateUrl = helpers_1.getHttpUrl(rollbackStep.stackTemplatePathOrUrl, this.deploymentBucket);
            const rollbackStackTemplatePath = helpers_1.getBucketKey(rollbackStep.stackTemplatePathOrUrl, this.deploymentBucket);
            this.deployment.push({
                deployment: null,
                rollback: {
                    ...rollbackStep,
                    stackTemplatePath: rollbackStackTemplatePath,
                    stackTemplateUrl: rollbackStackTemplateUrl,
                    region: this.region,
                    clientRequestToken: rollbackStep.clientRequestToken ? `rollback-${rollbackStep.clientRequestToken}` : undefined,
                },
            });
        };
        this.setPrinter = (printer) => {
            this.printer = printer;
        };
        this.startRollBackFn = async (_) => {
            var _a, _b;
            try {
                await ((_a = this.deploymentStateManager) === null || _a === void 0 ? void 0 : _a.startRollback());
                await ((_b = this.deploymentStateManager) === null || _b === void 0 ? void 0 : _b.startCurrentStep());
            }
            catch (err) {
                this.logger('startRolbackFn', [{ index: _.currentIndex }])(err);
            }
        };
        this.ensureStack = async (stackName) => {
            const result = await this.cfnClient.describeStacks({ StackName: stackName }).promise();
            return result.Stacks[0].StackStatus.endsWith('_COMPLETE');
        };
        this.ensureTemplateExists = async (templatePath) => {
            try {
                const bucketKey = helpers_1.getBucketKey(templatePath, this.deploymentBucket);
                await this.s3Client.headObject({ Bucket: this.deploymentBucket, Key: bucketKey }).promise();
                return true;
            }
            catch (e) {
                if (e.code === 'NotFound') {
                    throw new Error(`The cloudformation template ${templatePath} was not found in deployment bucket ${this.deploymentBucket}`);
                }
                this.logger('ensureTemplateExists', [{ templatePath }])(e);
                throw e;
            }
        };
        this.getTableStatus = async (tableName) => {
            var _a;
            assert_1.default(tableName, 'table name should be passed');
            try {
                const response = await this.ddbClient.describeTable({ TableName: tableName }).promise();
                const gsis = (_a = response.Table) === null || _a === void 0 ? void 0 : _a.GlobalSecondaryIndexes;
                return gsis ? gsis.every(idx => idx.IndexStatus === 'ACTIVE') : true;
            }
            catch (err) {
                this.logger('getTableStatus', [{ tableName }])(err);
                throw err;
            }
        };
        this.waitForActiveTables = async (tables) => {
            const throttledGetTableStatus = lodash_throttle_1.default(this.getTableStatus, this.options.throttleDelay);
            const waiters = tables.map(name => {
                return new Promise(resolve => {
                    let interval = setInterval(async () => {
                        const areIndexesReady = await throttledGetTableStatus(name);
                        if (areIndexesReady) {
                            clearInterval(interval);
                            resolve(undefined);
                        }
                    }, this.options.throttleDelay);
                });
            });
            await Promise.all(waiters);
        };
        this.waitForIndices = async (stackParams) => {
            var _a;
            if (stackParams.tableNames.length) {
                await new Promise(res => setTimeout(res, 2000));
            }
            await this.waitForActiveTables(stackParams.tableNames);
            try {
                await ((_a = this.deploymentStateManager) === null || _a === void 0 ? void 0 : _a.advanceStep());
            }
            catch (_b) {
            }
            return Promise.resolve();
        };
        this.preRollbackTableCheck = async (stackParams) => {
            await this.waitForActiveTables(stackParams.tableNames);
        };
        this.stackPollFn = (deploymentStep) => {
            let monitor;
            assert_1.default(deploymentStep.stackName, 'stack name should be passed to stackPollFn');
            if (this.printer) {
                monitor = new stack_event_monitor_1.StackEventMonitor(this.cfnClient, deploymentStep.stackName, this.printer);
                monitor.start();
            }
            return () => {
                if (monitor) {
                    monitor.stop();
                }
            };
        };
        this.doDeploy = async (currentStack) => {
            var _a;
            try {
                await ((_a = this.deploymentStateManager) === null || _a === void 0 ? void 0 : _a.startCurrentStep({ previousMetaKey: currentStack.previousMetaKey }));
            }
            catch (_b) {
            }
            const cfn = this.cfnClient;
            assert_1.default(currentStack.stackName, 'stack name should be passed to doDeploy');
            assert_1.default(currentStack.stackTemplateUrl, 'stackTemplateUrl must be passed to doDeploy');
            await this.ensureStack(currentStack.stackName);
            const parameters = Object.entries(currentStack.parameters).map(([key, val]) => {
                return {
                    ParameterKey: key,
                    ParameterValue: val.toString(),
                };
            });
            await cfn
                .updateStack({
                StackName: currentStack.stackName,
                Parameters: parameters,
                TemplateURL: currentStack.stackTemplateUrl,
                Capabilities: currentStack.capabilities,
                ClientRequestToken: currentStack.clientRequestToken,
            })
                .promise();
        };
        this.waitForDeployment = async (stackParams) => {
            const cfnClient = this.cfnClient;
            assert_1.default(stackParams.stackName, 'stackName should be passed to waitForDeployment');
            await cfnClient
                .waitFor('stackUpdateComplete', {
                StackName: stackParams.stackName,
            })
                .promise();
        };
        this.rollBackStack = async (currentStack) => {
            await this.doDeploy(currentStack);
        };
        this.logSpinnerMessage = (machine, message) => {
            this.logger(machine, [{ spinner: message }])();
            this.spinner.text = message;
        };
        this.options = {
            throttleDelay: 1000,
            eventPollingDelay: 1000,
            userAgent: undefined,
            ...options,
        };
        this.s3Client = new aws.S3(creds);
        this.cfnClient = new aws.CloudFormation({ ...creds, maxRetries: 10, customUserAgent: this.options.userAgent });
        this.ddbClient = new aws.DynamoDB({ ...creds, region, maxRetries: 10 });
        this.logger = aws_logger_1.fileLogger('deployment-manager');
    }
}
exports.DeploymentManager = DeploymentManager;
DeploymentManager.createInstance = async (context, deploymentBucket, spinner, options, printer) => {
    try {
        const cred = await configuration_manager_1.loadConfiguration(context);
        assert_1.default(cred.region);
        return new DeploymentManager(cred, cred.region, deploymentBucket, spinner, printer, options);
    }
    catch (e) {
        throw new Error('Could not load the credentials');
    }
};
