// amplify-nodejs-function-runtime-provider/lib/utils/legacyBuild.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildResource = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const amplify_function_plugin_interface_1 = require("amplify-function-plugin-interface");
const execa_1 = __importDefault(require("execa"));
const fs = __importStar(require("fs-extra"));
const glob_1 = __importDefault(require("glob"));
const path = __importStar(require("path"));
async function buildResource(request) {
    const resourceDir = request.service ? request.srcRoot : path.join(request.srcRoot, 'src');
    if (!request.lastBuildTimeStamp || isBuildStale(request.srcRoot, request.lastBuildTimeStamp, request.buildType, request.lastBuildType)) {
        installDependencies(resourceDir, request.buildType);
        if (request.legacyBuildHookParams) {
            runBuildScriptHook(request.legacyBuildHookParams.resourceName, request.legacyBuildHookParams.projectRoot);
        }
        return Promise.resolve({ rebuilt: true });
    }
    return Promise.resolve({ rebuilt: false });
}
exports.buildResource = buildResource;
function runBuildScriptHook(resourceName, projectRoot) {
    const scriptName = `amplify:${resourceName}`;
    if (scriptExists(projectRoot, scriptName)) {
        runPackageManager(projectRoot, undefined, scriptName);
    }
}
function scriptExists(projectRoot, scriptName) {
    const packageJsonPath = path.normalize(path.join(projectRoot, 'package.json'));
    if (fs.existsSync(packageJsonPath)) {
        const rootPackageJsonContents = require(packageJsonPath);
        return rootPackageJsonContents.scripts && rootPackageJsonContents.scripts[scriptName];
    }
    return false;
}
function installDependencies(resourceDir, buildType) {
    runPackageManager(resourceDir, buildType);
}
function runPackageManager(cwd, buildType, scriptName) {
    const packageManager = amplify_cli_core_1.getPackageManager(cwd);
    if (packageManager === null) {
        return;
    }
    const useYarn = packageManager.packageManager === 'yarn';
    const args = toPackageManagerArgs(useYarn, buildType, scriptName);
    try {
        execa_1.default.sync(packageManager.executable, args, {
            cwd,
            stdio: 'pipe',
            encoding: 'utf-8',
        });
    }
    catch (error) {
        if (error.code === 'ENOENT') {
            throw new Error(`Packaging lambda function failed. Could not find ${packageManager} executable in the PATH.`);
        }
        else {
            throw new Error(`Packaging lambda function failed with the error \n${error.message}`);
        }
    }
}
function toPackageManagerArgs(useYarn, buildType, scriptName) {
    if (scriptName) {
        return useYarn ? [scriptName] : ['run-script', scriptName];
    }
    const args = useYarn ? [] : ['install'];
    if (buildType === amplify_function_plugin_interface_1.BuildType.PROD) {
        args.push('--production');
    }
    return args;
}
function isBuildStale(resourceDir, lastBuildTimeStamp, buildType, lastBuildType) {
    const dirTime = new Date(fs.statSync(resourceDir).mtime);
    if (dirTime > lastBuildTimeStamp || buildType !== lastBuildType) {
        return true;
    }
    const fileUpdatedAfterLastBuild = glob_1.default
        .sync(`${resourceDir}/**`)
        .filter(p => !p.includes('dist'))
        .filter(p => !p.includes('node_modules'))
        .find(file => new Date(fs.statSync(file).mtime) > lastBuildTimeStamp);
    return !!fileUpdatedAfterLastBuild;
}
