// amplify-util-mock/lib/utils/lambda/populate-lambda-mock-env-vars.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.populateLambdaMockEnvVars = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const lodash_1 = __importDefault(require("lodash"));
const path = __importStar(require("path"));
const dotenv = __importStar(require("dotenv"));
const amplify_provider_awscloudformation_1 = require("amplify-provider-awscloudformation");
const populateLambdaMockEnvVars = async (context, processedLambda) => {
    processedLambda.environment = (await Promise.all([getAwsCredentials, getStaticDefaults, getDynamicDefaults, getDotEnvValues].map(envVarGetter => envVarGetter(processedLambda, context)))).reduce((acc, it) => ({ ...acc, ...it }), processedLambda.environment);
};
exports.populateLambdaMockEnvVars = populateLambdaMockEnvVars;
const getAwsCredentials = async (_, context) => {
    const env = amplify_cli_core_1.stateManager.getLocalEnvInfo().envName;
    let appId = undefined;
    try {
        appId = amplify_provider_awscloudformation_1.resolveAppId(context);
    }
    catch (_a) {
    }
    const awsConfigInfo = await amplify_provider_awscloudformation_1.loadConfigurationForEnv(context, env, appId);
    return {
        AWS_ACCESS_KEY_ID: awsConfigInfo.accessKeyId,
        AWS_SECRET_ACCESS_KEY: awsConfigInfo.secretAccessKey,
        AWS_SESSION_TOKEN: awsConfigInfo.sessionToken,
    };
};
const getStaticDefaults = () => ({
    _X_AMZN_TRACE_ID: 'amplify-mock-x-amzn-trace-id',
    AWS_EXECUTION_ENV: 'AWS_Lambda_amplify-mock',
    AWS_LAMBDA_FUNCTION_MEMORY_SIZE: '128',
    AWS_LAMBDA_FUNCTION_VERSION: '1',
    AWS_LAMBDA_INITIALIZATION_TYPE: 'on-demand',
    AWS_LAMBDA_LOG_GROUP_NAME: 'amplify-mock-aws-lambda-log-group-name',
    AWS_LAMBDA_LOG_GROUP_STREAM_NAME: 'amplify-mock-aws-lambda-log-group-stream-name',
    TZ: 'UTC',
});
const getDynamicDefaults = (processedLambda) => {
    const env = amplify_cli_core_1.stateManager.getLocalEnvInfo().envName;
    const teamProvider = amplify_cli_core_1.stateManager.getTeamProviderInfo();
    const region = lodash_1.default.get(teamProvider, [env, 'awscloudformation', 'Region']);
    const lambdaPath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), 'function', processedLambda.name);
    return {
        _HANDLER: processedLambda.handler,
        AWS_REGION: region,
        AWS_LAMBDA_FUNCTION_NAME: processedLambda.name,
        LAMBDA_TASK_ROOT: lambdaPath,
        LAMBDA_RUNTIME_DIR: lambdaPath,
    };
};
const getDotEnvValues = (processedLambda) => {
    try {
        const result = dotenv.config({ path: path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), 'function', processedLambda.name, '.env') });
        if (result.error) {
            throw result.error;
        }
        return result.parsed;
    }
    catch (_a) {
    }
};
