// @aws-amplify/graphql-transformer-core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappingTemplate = exports.getAppSyncServiceExtraDirectives = exports.TransformerPluginBase = exports.TransformerModelEnhancerBase = exports.TransformerModelBase = exports.DirectiveWrapper = exports.collectDirectivesByTypeNames = exports.collectDirectives = exports.ConflictHandlerType = exports.validateModelSchema = exports.SyncUtils = exports.GraphQLTransform = void 0;
const graphql_1 = require("graphql");
const validation_1 = require("./transformation/validation");
var transformation_1 = require("./transformation");
Object.defineProperty(exports, "GraphQLTransform", { enumerable: true, get: function () { return transformation_1.GraphQLTransform; } });
Object.defineProperty(exports, "SyncUtils", { enumerable: true, get: function () { return transformation_1.SyncUtils; } });
var validation_2 = require("./transformation/validation");
Object.defineProperty(exports, "validateModelSchema", { enumerable: true, get: function () { return validation_2.validateModelSchema; } });
var index_1 = require("./config/index");
Object.defineProperty(exports, "ConflictHandlerType", { enumerable: true, get: function () { return index_1.ConflictHandlerType; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "collectDirectives", { enumerable: true, get: function () { return utils_1.collectDirectives; } });
Object.defineProperty(exports, "collectDirectivesByTypeNames", { enumerable: true, get: function () { return utils_1.collectDirectivesByTypeNames; } });
Object.defineProperty(exports, "DirectiveWrapper", { enumerable: true, get: function () { return utils_1.DirectiveWrapper; } });
__exportStar(require("./errors"), exports);
var transformer_plugin_base_1 = require("./transformation/transformer-plugin-base");
Object.defineProperty(exports, "TransformerModelBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerModelBase; } });
Object.defineProperty(exports, "TransformerModelEnhancerBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerModelEnhancerBase; } });
Object.defineProperty(exports, "TransformerPluginBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerPluginBase; } });
const getAppSyncServiceExtraDirectives = () => {
    return graphql_1.print(validation_1.EXTRA_DIRECTIVES_DOCUMENT);
};
exports.getAppSyncServiceExtraDirectives = getAppSyncServiceExtraDirectives;
var cdk_compat_1 = require("./cdk-compat");
Object.defineProperty(exports, "MappingTemplate", { enumerable: true, get: function () { return cdk_compat_1.MappingTemplate; } });
