// @aws-amplify/graphql-transformer-core/lib/transformer-context/resource-helper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformerResourceHelper = void 0;
const md5_1 = __importDefault(require("md5"));
class TransformerResourceHelper {
    constructor(stackManager) {
        this.stackManager = stackManager;
        this.generateResourceName = (name) => {
            if (!this.api) {
                throw new Error('API not initialized');
            }
            this.ensureEnv();
            const env = this.stackManager.getParameter('env').valueAsString;
            const apiId = this.api.apiId;
            return `${name}-${apiId}-${env}`;
        };
        this.generateIAMRoleName = (name) => {
            if (!this.api) {
                throw new Error('API not initialized');
            }
            this.ensureEnv();
            const env = this.stackManager.getParameter('env').valueAsString;
            const apiId = this.api.apiId;
            const shortName = name.slice(0, 64 - 38 - 6) + md5_1.default(name).slice(0, 6);
            return `${shortName}-${apiId}-${env}`;
        };
        this.ensureEnv = () => {
            if (!this.stackManager.getParameter('env')) {
                this.stackManager.addParameter('env', {
                    type: 'String',
                    default: 'NONE',
                });
            }
        };
    }
    bind(api) {
        this.api = api;
    }
}
exports.TransformerResourceHelper = TransformerResourceHelper;
