// amplify-provider-awscloudformation/lib/graphql-transformer/transform-config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readSchema = exports.throwIfNotJSONExt = exports.loadProject = exports.writeConfig = exports.loadConfig = exports.TRANSFORM_CONFIG_FILE_NAME = void 0;
const path = __importStar(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
exports.TRANSFORM_CONFIG_FILE_NAME = `transform.conf.json`;
async function loadConfig(projectDir) {
    let config = {};
    try {
        const configPath = path.join(projectDir, exports.TRANSFORM_CONFIG_FILE_NAME);
        const configExists = fs_extra_1.default.existsSync(configPath);
        if (configExists) {
            const configStr = await fs_extra_1.default.readFile(configPath, 'utf-8');
            config = JSON.parse(configStr);
        }
        return config;
    }
    catch (err) {
        return config;
    }
}
exports.loadConfig = loadConfig;
async function writeConfig(projectDir, config) {
    const configFilePath = path.join(projectDir, exports.TRANSFORM_CONFIG_FILE_NAME);
    await fs_extra_1.default.writeFile(configFilePath, JSON.stringify(config, null, 4));
    return config;
}
exports.writeConfig = writeConfig;
async function loadProject(projectDirectory, opts) {
    const schema = await readSchema(projectDirectory);
    const functions = {};
    if (!(opts && opts.disableFunctionOverrides === true)) {
        const functionDirectory = path.join(projectDirectory, 'functions');
        const functionDirectoryExists = fs_extra_1.default.existsSync(functionDirectory);
        if (functionDirectoryExists) {
            const functionFiles = await fs_extra_1.default.readdir(functionDirectory);
            for (const functionFile of functionFiles) {
                if (functionFile.indexOf('.') === 0) {
                    continue;
                }
                const functionFilePath = path.join(functionDirectory, functionFile);
                functions[functionFile] = functionFilePath;
            }
        }
    }
    const pipelineFunctions = {};
    if (!(opts && opts.disablePipelineFunctionOverrides === true)) {
        const pipelineFunctionDirectory = path.join(projectDirectory, 'pipelineFunctions');
        const pipelineFunctionDirectoryExists = fs_extra_1.default.existsSync(pipelineFunctionDirectory);
        if (pipelineFunctionDirectoryExists) {
            const pipelineFunctionFiles = await fs_extra_1.default.readdir(pipelineFunctionDirectory);
            for (const pipelineFunctionFile of pipelineFunctionFiles) {
                if (pipelineFunctionFile.indexOf('.') === 0) {
                    continue;
                }
                const pipelineFunctionPath = path.join(pipelineFunctionDirectory, pipelineFunctionFile);
                pipelineFunctions[pipelineFunctionFile] = await fs_extra_1.default.readFile(pipelineFunctionPath);
            }
        }
    }
    const resolvers = {};
    if (!(opts && opts.disableResolverOverrides === true)) {
        const resolverDirectory = path.join(projectDirectory, 'resolvers');
        const resolverDirExists = fs_extra_1.default.existsSync(resolverDirectory);
        if (resolverDirExists) {
            const resolverFiles = await fs_extra_1.default.readdir(resolverDirectory);
            for (const resolverFile of resolverFiles) {
                if (resolverFile.indexOf('.') === 0) {
                    continue;
                }
                const resolverFilePath = path.join(resolverDirectory, resolverFile);
                resolvers[resolverFile] = await fs_extra_1.default.readFile(resolverFilePath);
            }
        }
    }
    const stacksDirectory = path.join(projectDirectory, 'stacks');
    const stacksDirExists = fs_extra_1.default.existsSync(stacksDirectory);
    const stacks = {};
    if (stacksDirExists) {
        const stackFiles = await fs_extra_1.default.readdir(stacksDirectory);
        for (const stackFile of stackFiles) {
            if (stackFile.indexOf('.') === 0) {
                continue;
            }
            const stackFilePath = path.join(stacksDirectory, stackFile);
            throwIfNotJSONExt(stackFile);
            const stackBuffer = await fs_extra_1.default.readFile(stackFilePath);
            try {
                stacks[stackFile] = JSON.parse(stackBuffer.toString());
            }
            catch (e) {
                throw new Error(`The CloudFormation template ${stackFiles} does not contain valid JSON.`);
            }
        }
    }
    const config = await loadConfig(projectDirectory);
    return {
        functions,
        pipelineFunctions,
        stacks,
        resolvers,
        schema,
        config,
    };
}
exports.loadProject = loadProject;
function throwIfNotJSONExt(stackFile) {
    const extension = path.extname(stackFile);
    if (extension === '.yaml' || extension === '.yml') {
        throw new Error(`Yaml is not yet supported. Please convert the CloudFormation stack ${stackFile} to json.`);
    }
    if (extension !== '.json') {
        throw new Error(`Invalid extension ${extension} for stack ${stackFile}`);
    }
}
exports.throwIfNotJSONExt = throwIfNotJSONExt;
async function readSchema(projectDirectory) {
    const schemaFilePath = path.join(projectDirectory, 'schema.graphql');
    const schemaDirectoryPath = path.join(projectDirectory, 'schema');
    const schemaFileExists = fs_extra_1.default.existsSync(schemaFilePath);
    const schemaDirectoryExists = fs_extra_1.default.existsSync(schemaDirectoryPath);
    let schema;
    if (schemaFileExists) {
        schema = (await fs_extra_1.default.readFile(schemaFilePath)).toString();
    }
    else if (schemaDirectoryExists) {
        schema = (await readSchemaDocuments(schemaDirectoryPath)).join('\n');
    }
    else {
        throw new Error(`Could not find a schema at ${schemaFilePath}`);
    }
    return schema;
}
exports.readSchema = readSchema;
async function readSchemaDocuments(schemaDirectoryPath) {
    const files = await fs_extra_1.default.readdir(schemaDirectoryPath);
    let schemaDocuments = [];
    for (const fileName of files) {
        if (fileName.indexOf('.') === 0) {
            continue;
        }
        const fullPath = `${schemaDirectoryPath}/${fileName}`;
        const stats = await fs_extra_1.default.lstat(fullPath);
        if (stats.isDirectory()) {
            const childDocs = await readSchemaDocuments(fullPath);
            schemaDocuments = schemaDocuments.concat(childDocs);
        }
        else if (stats.isFile()) {
            const schemaDoc = await fs_extra_1.default.readFile(fullPath);
            schemaDocuments.push(schemaDoc);
        }
    }
    return schemaDocuments;
}
