// amplify-util-headless-input/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateImportAuthRequest = exports.validateUpdateAuthRequest = exports.validateAddAuthRequest = exports.validateUpdateApiRequest = exports.validateAddApiRequest = exports.validateAddStorageRequest = void 0;
const HeadlessInputValidator_1 = require("./HeadlessInputValidator");
const schemaSuppliers_1 = require("./schemaSuppliers");
const upgradePipelines_1 = require("./upgradePipelines");
const validateAddStorageRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.addStorageRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateAddStorageRequest = validateAddStorageRequest;
const validateAddApiRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.addApiRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateAddApiRequest = validateAddApiRequest;
const validateUpdateApiRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.updateApiRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateUpdateApiRequest = validateUpdateApiRequest;
const validateAddAuthRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.addAuthRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateAddAuthRequest = validateAddAuthRequest;
const validateUpdateAuthRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.updateAuthRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateUpdateAuthRequest = validateUpdateAuthRequest;
const validateImportAuthRequest = (raw) => {
    return new HeadlessInputValidator_1.HeadlessInputValidator(schemaSuppliers_1.importAuthRequestSchemaSupplier, upgradePipelines_1.noopUpgradePipeline).validate(raw);
};
exports.validateImportAuthRequest = validateImportAuthRequest;
