// amplify-util-headless-input/lib/schemaSuppliers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateApiRequestSchemaSupplier = exports.addApiRequestSchemaSupplier = exports.importAuthRequestSchemaSupplier = exports.updateAuthRequestSchemaSupplier = exports.addAuthRequestSchemaSupplier = exports.addStorageRequestSchemaSupplier = void 0;
const addStorageRequestSchemaSupplier = version => {
    return getSchema('AddStorageRequest', 'storage', version);
};
exports.addStorageRequestSchemaSupplier = addStorageRequestSchemaSupplier;
const addAuthRequestSchemaSupplier = version => {
    return getSchema('AddAuthRequest', 'auth', version);
};
exports.addAuthRequestSchemaSupplier = addAuthRequestSchemaSupplier;
const updateAuthRequestSchemaSupplier = version => {
    return getSchema('UpdateAuthRequest', 'auth', version);
};
exports.updateAuthRequestSchemaSupplier = updateAuthRequestSchemaSupplier;
const importAuthRequestSchemaSupplier = version => {
    return getSchema('ImportAuthRequest', 'auth', version);
};
exports.importAuthRequestSchemaSupplier = importAuthRequestSchemaSupplier;
const addApiRequestSchemaSupplier = version => {
    return getSchema('AddApiRequest', 'api', version);
};
exports.addApiRequestSchemaSupplier = addApiRequestSchemaSupplier;
const updateApiRequestSchemaSupplier = version => {
    return getSchema('UpdateApiRequest', 'api', version);
};
exports.updateApiRequestSchemaSupplier = updateApiRequestSchemaSupplier;
const getSchema = async (type, category, version) => {
    try {
        return {
            rootSchema: await Promise.resolve().then(() => __importStar(require(`amplify-headless-interface/schemas/${category}/${version}/${type}.schema.json`))),
        };
    }
    catch (ex) {
        return;
    }
};
